Redundant streams fallback behavior for alternate renditions with multiple media groups

tchakabam opened the pull request:
### Description of the Changes

Until now, we are not supporting a typical live broadcast use-case: using backup streams for all alternate audio and subtitle rendition options, grouping these into respective group-IDs. For example group-ID A groups all the primary renditions and video-levels, B all the back ones. In this case, group-IDs of selected audio and subtitle renditions need to be matched to the group-ID of the running level and vice versa. So if we fallback on a level, we have to re-select the rendition so to match the group ID, and if we fallback on audio we may have to switch to the matching redundant level. That is because the grouped renditions/levels actually run on sync'ed encoders, while if the groups don't match, there can easily be A/V-sync issues. In fact, the clocks of the redundants streams across group IDs are not actually even specified to have anything like matching clocks or so, they could operate in completely different timeline.

**=> This PR will fix the above and allow proper fallback behavior across all directions (video-levels <-> audio-and-subs-renditions)**

### CheckLists

- [x] changes have been done against master branch, and PR does not conflict
- [x] no commits have been done in dist folder (we will take care of updating it)
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [x] API or design changes are documented in API.md

tchakabam commented:
May fix parts of https://github.com/video-dev/hls.js/issues/1438

netTrekfd commented on src/loader/playlist-loader.js:
```diff
@@ -152,24 +152,26 @@ class PlaylistLoader extends EventHandler {
 
   onManifestLoading (data) {
-    this.load(data.url, { type: ContextType.MANIFEST });
+    this.load(data.url, { type: ContextType.MANIFES, level: null, id: null });
```
There is a little type error, should be:
this.load(data.url, { type: ContextType.MANIFEST, level: null, id: null });

tchakabam replied:
which is why stuff currently fails :+1:

tchakabam commented on src/controller/audio-track-controller.js:
```diff
@@ -81,73 +80,130 @@ class AudioTrackController extends EventHandler {
 
-  /** select an audio track, based on its index in audio track lists**/
-  set audioTrack (audioTrackId) {
-    if (this.trackId !== audioTrackId || this.tracks[audioTrackId].details === undefined)
```
@mangui how did the condition to run the set internal make sense: `this.tracks[audioTrackId].details === undefined`

wouldn't you still want to switch even if the metadata is not loaded yet?

tchakabam replied:
see https://github.com/video-dev/hls.js/commit/2284bc65840aa31c544c4a1c3fbb6c0068500134#diff-b5ce37fcb02b92e3f287ca1a1e3ad800R65

mangui replied:
condition is : 
call `set internal` if new track id OR if same track id but details not available yet (in that case it will force a reload of the details)

tchakabam replied:
of course, makes sense :) thanks for the explanation. i have added the check for the `details` prop in this branch now as well to make sure we are allowing to switch track in this case as well.

tchakabam commented:
There is another error in here though, related to audio-tracks switching, investigating right now

tchakabam commented:
@mangui @johnBartos @tjenkinson Hey guys, do you think you could check this out? just to make sure i didn't break something badly :) 

Will also add unit tests for this behavior hopefully as soon as I get to it!

mangui commented on src/controller/abr-controller.js:
```diff
@@ -31,7 +31,12 @@ class AbrController extends EventHandler {
 
   onFragLoading (data) {
-    let frag = data.frag;
+    const hls = this.hls;
```
`hls` and `config` objects were only initialized if useful before.
now you will have extra processing everytime a subtitle or audio frag is loading.
you could argue that this might look like micro-optimization, but still ;-)

mangui commented on src/controller/abr-controller.js:
```diff
@@ -40,12 +45,9 @@ class AbrController extends EventHandler {
-          isLive = hls.levels[level].details.live,
-          config = hls.config,
-          ewmaFast, ewmaSlow;
+        let level = data.frag.level;
```
`let level = frag.level` is shorter

mangui commented on src/controller/abr-controller.js:
```diff
@@ -55,8 +57,9 @@ class AbrController extends EventHandler {
         this._bwEstimator = new EwmaBandWidthEstimator(hls, ewmaSlow, ewmaFast, config.abrEwmaDefaultEstimate);
       }
-      this.fragCurrent = frag;
+      this._bwEstimator = new EwmaBandWidthEstimator(hls, ewmaSlow, ewmaFast, config.abrEwmaDefaultEstimate);
```
looks like bwEstimator is created twice now ?

tchakabam replied:
true. was debugging this stuff and left some mess around :/

mangui commented on src/controller/abr-controller.js:
```diff
@@ -55,8 +57,9 @@ class AbrController extends EventHandler {
-      this.fragCurrent = frag;
+      this._bwEstimator = new EwmaBandWidthEstimator(hls, ewmaSlow, ewmaFast, config.abrEwmaDefaultEstimate);
     }
+    this.fragCurrent = frag;
```
we should not update `this.fragCurrent` if `frag` is not from `main`

tchakabam replied:
i was actually wondering about that. i think there was a bug however here, we were sometimes hitting an undefined `fragCurrent`..? but maybe it's fixed now, i was trying to debug this part and something was fishy.

axten commented:
Hi @tchakabam,
can you confirm that with this great PR a master playlist with multiple media groups (GROUP-ID) for video (with subs) and audio is handled correctly?
https://zdfhlstest2-i.akamaihd.net/hls/live/226660/zdf_ahw/high/master.m3u8

And if yes, does hls.js ensure that all active medias are from the same group?
**Example:** video, audio and subs from group A are playing. Now video fails and hls.js switches to fallback video stream from group B. Does hls.js switch also audio and subtitle streams to Group B in this case?

tchakabam commented:
Hi @axten Yes. That's exactly the idea.

axten commented:
ok great! this is exactly what we need! cannot await this to land :)

tchakabam commented:
@axten That's what I had been told as well 👍 Very well, then this will need some clean-up and polishing, but the functionality is now there. We were actually missing the information inside our Quality-Level data-model where we are handling redundant streams by nesting various URLs for the same quality, but as a result were compressing away the various audio-group IDs for the respective redundants streams. This information is now being patched onto the data-model along with the redundant URLs. 

Furthermore, the audio-track switched event is being handled now by the level-controller in order to make sure that the current level played is actually for the group ID of the latest selected audio-track. If that is not the case, we will correct that setting internally and re-load the video levels accordingly. 

In a way like a mirror, the analogous logic is implemented for audio-tracks. As a quality level is switched (wether it be for any reason, manually or by recovery/fallback mechanisms) we check which actual stream is being played of the redundant options, and which audio-group-id that corresponds. If that doesn't match the audio-group-ID of the currently selected audio-track, we switch it accordingly, while trying to preserve the actual name/language setting which indicates the logical "role" of this audio track (to avoid switching say from original audio to audio for visibility handicap etc).

Further things to do now would be:

- Add the analoguous logic for subtitles redundant groups (will be pretty much a copy of what we added for audio-tracks, so very easy)

- Add some unit tests to spec this behavior clearly and avoid regression

- Add these test cases of yours to our acceptance test suite 

=> Then we should be able to merge this into our master branch.

You could anyhow already use this branch here for your testing to validate we got it all right, and in principle also for your prod player already, if it works for you (and after we removed a few hard-coded log lines maybe :)).

walkerflocker commented on src/controller/audio-track-controller.js:
```diff
@@ -1,156 +1,330 @@
+      }
+      const newTrack = this.tracks[i];
+      if (newTrack.name === name &&
+       newTrack.language === language) {
```
Why check for both `name` and `language` here but only `name` on [_selectInitialAudioTrack](https://github.com/video-dev/hls.js/blob/feature/audio-track-fallback/src/controller/audio-track-controller.js#L220)? From what I see in the spec only `name` is required to both exist and be consistent across groups. Any sane manifest will have matching and unique names and languages, but it may be worth being consistent

tchakabam replied:
Good point. Name should be sufficient, you're right.

tchakabam commented:
Build pipeline succeeded (not sure why it's greyed out however): https://travis-ci.org/video-dev/hls.js/builds/389764334
